torchsparse/backend/convolution/convolution_cuda.cu

```diff
@@ -52,8 +268,410 @@ __global__ void scatter_kernel(const int n_in, const int n_out, const int c,
+  for (int i = 0; i < split_points.size(); i++) {
+    int cur_split_point = split_points[i];
+    int prev_split_point = i > 0 ? split_points[i - 1] : 0;
+    std::vector<int> cur_group;
+    group_max_size = -1;
+    for (int j = prev_split_point; j < cur_split_point; j++) {
+      cur_group.push_back(kernel_order[j]);
+      group_max_size = std::max(
+          group_max_size, *(neighbor_offset.data_ptr<int>() + kernel_order[j]));
+    }
+    groups.push_back(cur_group);
+    group_sizes.push_back(group_max_size);
+    if (group_max_size < mm_thresh) {
+      // bmm
+      mm_ops.push_back(1);
+    } else {
+      // separate
+      mm_ops.push_back(0);
+    }
+  }
+
+  // determine the cumulated buffer size for each offset
+
+  int cur_cum_size = 0;
+  for (int i = 0; i < groups.size(); i++) {
+    for (int j = 0; j < groups[i].size(); j++) {
+      *(cum_buffer_sizes.data_ptr<int>() + groups[i][j]) = cur_cum_size;
+      if (mm_ops[i] == 0) {
+        cur_cum_size += *(neighbor_offset.data_ptr<int>() + groups[i][j]);
+        buffer_size += *(neighbor_offset.data_ptr<int>() + groups[i][j]);
+      } else {
+        cur_cum_size += group_sizes[i];
+        buffer_size += group_sizes[i];
+      }
+    }
+  }
+}
+
+void convolution_forward_cuda_latest(
+    at::Tensor in_feat, at::Tensor out_feat, at::Tensor _kernel,
+    at::Tensor neighbor_map, at::Tensor neighbor_offset, at::Tensor input_mask,
+    at::Tensor output_mask, const float epsilon, const int mm_thresh,
+    const int conv_mode, const bool transpose, at::Tensor global_buffer) {
+  if (in_feat.size(1) != _kernel.size(1)) {
+    throw std::invalid_argument("Input feature size and kernel size mismatch");
+  }
+
+  at::Tensor neighbor_offset_gpu = neighbor_offset.to(in_feat.device());
+  at::Tensor neighbor_offset_cum_gpu =
+      torch::cumsum(neighbor_offset_gpu, 0).to(at::ScalarType::Int);
+  at::Tensor neighbor_offset_cum =
+      neighbor_offset_cum_gpu.to(neighbor_offset.device());
+  // at::Tensor neighbor_map_cpu = neighbor_map.to(neighbor_offset.device());
+  at::Tensor cum_buffer_sizes = torch::zeros_like(neighbor_offset);
+
+  // int tile_size = 1024;
+  int tile_id;
+
+  auto options =
+      torch::TensorOptions().dtype(in_feat.dtype()).device(in_feat.device());
+  bool is_half = in_feat.scalar_type() == at::ScalarType::Half;
+
+  // pad num channels to an even number
+  at::Tensor kernel = _kernel.clone();
+  if (is_half) {
+    if (in_feat.size(1) % 8 != 0) {
+      in_feat = torch::cat(
+          {in_feat,
+           torch::zeros({in_feat.size(0), 8 - (in_feat.size(1) % 8)}, options)},
+          -1);
+      kernel = torch::cat(
+          {kernel, torch::zeros({kernel.size(0), 8 - (kernel.size(1) % 8),
+                                 kernel.size(2)},
+                                options)},
+          1);
+    }
+    if (out_feat.size(1) % 8 != 0) {
+      out_feat = torch::cat(
+          {out_feat, torch::zeros({out_feat.size(0), 8 - (in_feat.size(1) % 8)},
```

Comment:
The format of line 441-470 is a little bit messy. Is it caused by the linting tool?

Reply:
Yes. The clang-format seems to be pretty insistent on this formating.